File: content/snippets/cpp/interview_lcci/04/04_check_balance.cpp
#include <iostream>

struct Node { int v; Node* l; Node* r; Node(int x): v(x), l(nullptr), r(nullptr) {} };

int heightOrFail(Node* n) {
    if (!n) return 0;
    int lh = heightOrFail(n->l); if (lh == -1) return -1;
    int rh = heightOrFail(n->r); if (rh == -1) return -1;
    if (lh - rh > 1 || rh - lh > 1) return -1;
    return 1 + (lh > rh ? lh : rh);
}

bool isBalanced(Node* root) { return heightOrFail(root) != -1; }

int main() {
    Node* a = new Node(1); a->l = new Node(2); a->r = new Node(3);
    Node* b = new Node(1); b->l = new Node(2); b->l->l = new Node(3); b->l->l->l = new Node(4);
    std::cout << std::boolalpha << isBalanced(a) << '\n';
    std::cout << std::boolalpha << isBalanced(b) << '\n';
}

File: content/snippets/scala2/interview_lcci/04/04_check_balance.scala
object Lcci0404CheckBalance {
  final case class Node(value: Int, var left: Node | Null = null, var right: Node | Null = null)

  def heightOrFail(n: Node | Null): Int = {
    if (n == null) return 0
    val lh = heightOrFail(n.asInstanceOf[Node].left)
    if (lh == -1) return -1
    val rh = heightOrFail(n.asInstanceOf[Node].right)
    if (rh == -1) return -1
    if (math.abs(lh - rh) > 1) return -1
    1 + math.max(lh, rh)
  }

  def isBalanced(root: Node | Null): Boolean = heightOrFail(root) != -1

  def main(args: Array[String]): Unit = {
    val a = Node(1, Node(2), Node(3))
    val b = Node(1, Node(2, Node(3, Node(4), null), null), null)
    println(isBalanced(a))
    println(isBalanced(b))
  }
}
